File: topkElements/topkElements.py
# Sorting Solution

class Solution:
    def topKFrequent(self, nums: list[int], k: int) -> list[int]:
        freqs = {}
        for num in nums:
            freqs[num] = freqs.get(num, 0) + 1 # increment entry
        
        sorted_by_value = dict(sorted(freqs.items(), key= lambda item: item[1],reverse=True))
        return list(sorted_by_value)[:k]
    
# Min Heap Solution
import heapq
class Solution:
    def topKFrequent(self, nums: list[int], k: int) -> list[int]:
        freqs = {}
        for num in nums:
            freqs[num] = freqs.get(num, 0) + 1 # increment entry
        
        heap = [] 
        for num,count in freqs.items():
            heapq.heappush(heap,(count,num)) # keeps the minimimum frequency on top of the heap

            if len(heap) > k:
                heapq.heappop(heap)
        
        res = []
        for i in range(k):
            res.append(heapq.heappop(heap)[1])
        return res

# Bucket sort solution

class Solution:
    def topKFrequent(self, nums: list[int], k: int) -> list[int]:
        freqs = {} # Dict of frequency counts
        buckets = [[] for i in range(len(nums))] # buckets for each frequence; Each index corresponds to the the frequency and will be appended with the 
                                                # num of matching frequency.
        for num in nums:
            freqs[num] = freqs.get(num, 0) + 1 # increment entry

        for num,count in freqs.items():
            buckets[count-1].append(num)
        
        res = []
        for lst in buckets[::-1]:
            for num in lst:
                res.append(num)
                if len(res) == k:
                    return res
        
        

File: topkElements/topkElements.cpp
// Heap solution

#include <unordered_map>
#include <vector>
#include <queue>
#include <utility>
#include <functional>

class Solution {
public:
    std::vector<int> topKFrequent(std::vector<int>& nums, int k) {
        std::unordered_map<int,int> freqs;

        for (int i = 0; i < nums.size();i++){
            freqs[nums[i]]++; // Get the counts for each number
        }

        std::priority_queue<std::pair<int,int>, std::vector<std::pair<int,int>>, std::greater<std::pair<int,int>>> heap;

        for (auto& entry : freqs){
            heap.push({entry.second,entry.first});
            if (heap.size() > k){
                heap.pop();
            }
        }

        std::vector<int> res;

        for (int i = 0; i < k; i++){
            res.push_back(heap.top().second); // Save the top entry
            heap.pop(); // Remove the top 
        }
        
        return res;
    }


};

// Bucket sort solution

class Solution2 {
public:
    std::vector<int> topKFrequent(std::vector<int>& nums, int k) {
        std::unordered_map<int,int> freqs;
        std::vector<std::vector<int>> buckets(nums.size() + 1);

        for (int i = 0; i < nums.size();i++){
            freqs[nums[i]]++; // Get the counts for each number
        }

        for (const auto& entry : freqs){
            buckets[entry.second].push_back(entry.first);
        }

        std::vector<int> res;
        for (int i = buckets.size() - 1; i > 0; --i) {
            for (int n : buckets[i]) {
                res.push_back(n);
                if (res.size() == k) {
                    return res;
                }
            }
        }
        return res;
    }
};